### PR TITLE
Interpolate tensor operation (Inference Only) 

### LINKED ISSUE #455: Implement interpolate function (inference only)
### Feature description

Implement [interpolate](https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html?highlight=interpolate#torch.nn.functional.interpolate)

### Feature motivation

Interpolation is helpful for downsampling/upsampling the feature maps

### PR DESCRIPTION
**Pull request by Nikaidou-Shinku**
### Checklist

- [ ] Confirmed that `run-checks all` script has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Fixes #455

### Changes

- ✅: This feature is complete.
- 🚧: I'm working on this feature.
- ⚠️: This feature requires upstream support.
- 🆘: I need help completing this feature.

|Backend\Feature|Nearest-neighbor|Bilinear|Bicubic|Anti-aliasing|
|:-:|:-:|:-:|:-:|:-:|
|WGPU|✅|✅|✅|🚧|
|Candle|✅|⚠️|⚠️|⚠️|
|LibTorch|✅|✅|✅|🚧|
|NdArray|✅|✅|✅|🚧|
|Autodiff|🆘|🆘|🆘|🆘|
|Fusion|✅|✅|✅|🚧|

### Notes

- Candle currently only supports nearest-neighbor interpolation, and there is an issue about adding bilinear interpolation: huggingface/candle#1392.
- I haven't learned much about ML, so I may need help from others with the Autodiff backend. If you are interested in these features, come and help implement them!
- With interpolation support, we can implement ONNX's Resize operator in Burn, there is an issue about it: #510. Maybe we can implement it in this PR, or in a new PR after this PR is merged.
- If you have any other needs related to interpolation, welcome to discuss!

### Testing

I will write new test cases to test these features, but currently the test cases are lacking because many features are not completed yet.

### Completed

- [x] Nearest-neighbor interpolation for WGPU, Candle, LibTorch, NdArray and Fusion backends.
- [x] Bilinear interpolation for WGPU, LibTorch, NdArray and Fusion backends.
- [x] Bicubic interpolation for WGPU, LibTorch, NdArray and Fusion backends.

All these implementations give results consistent with PyTorch.

### Not done

- [ ] Autodiff backend. (which means that training is currently unavailable) (#1393)
- [ ] Anti-aliasing for bilinear and bicubic interpolation. (#1393)
- [ ] Parameter `align_corners` for bilinear and bicubic interpolation. (for more information check [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html#torch.nn.functional.interpolate))  (#1393)
- [ ] Mode `nearest-exact`. (see the link previous for details)  (#1393)
- [ ] ONNX's Resize operator. (#455)
- [ ] More test cases.  (#1393)
- [ ] Documentation  (#1393)

### CODE REVIEW

**Comment by Nikaidou-Shinku** [comment]
> I have some idea on how to implement the Autodiff for this operation and would like to help out on this.

Hello! I'm honestly at a loss as to how to implement autodiff, so it would be great if you could help!

> Let me know If there have been any new updates.

I'm sorry that I'm a little busy during this time and may not be able to push this PR forward. But I will find a way to get it done before March, and I really hope to see this new feature in the next big version of Burn!

**Comment by Nikaidou-Shinku** [comment]
Sorry I may not have time to continue working on this PR at the moment, I would be grateful if someone could take over it, or maybe consider merging the already completed parts and keeping the rest as TODO.

I will list the current progress of this PR below.

## Completed

- [x] Nearest-neighbor interpolation for WGPU, Candle, LibTorch, NdArray and Fusion backends.
- [x] Bilinear interpolation for WGPU, LibTorch, NdArray and Fusion backends.
- [x] Bicubic interpolation for WGPU, LibTorch, NdArray and Fusion backends.

All these implementations give results consistent with PyTorch.

## Not done

- [ ] Autodiff backend. (which means that training is currently unavailable)
- [ ] Anti-aliasing for bilinear and bicubic interpolation.
- [ ] Parameter `align_corners` for bilinear and bicubic interpolation. (for more information check [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html#torch.nn.functional.interpolate))
- [ ] ONNX's Resize operator.
- [ ] More test cases.

**Review by antimora — APPROVED** [review]
@Nikaidou-Shinku , thank you so much for this contribution! This is a big step for the full feature. We can start working on the Resize ONNX op which was a major missing op for many image related ONNX could not be implemented.

It is very responsible of you to list of TODOs and what it is completed. 

I am approving it so it could be merged. We can file all remaining TODOs into an issue.

@ashdtu , @louisfd Please reviewing it and let's merge it.

File: crates/burn-tensor/src/tensor/ops/modules/base.rs [hunk 1]
```diff
@@ -114,6 +114,28 @@ pub struct UnfoldOptions {
...
+    Bilinear,
+
+    /// Bicubic interpolation.
+    /// <https://en.wikipedia.org/wiki/Bicubic_interpolation>
+    Bicubic,
+}
+
+/// Interpolation options.
+#[derive(new)]
+pub struct InterpolateOptions {
```
Comment (ashdtu) [inline crates/burn-tensor/src/tensor/ops/modules/base.rs:134]: I think this is great for future proofing. later on we can add other options like align corners here.

File: crates/burn-ndarray/src/ops/interpolate.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    run_par!(|| {
+        iter_range_par!(0, out_element_num).for_each(|id| {
+            let (b, c, h, w) = (
+                id / strides.0,
+                id % strides.0 / strides.1,
+                id % strides.1 / strides.2,
+                id % strides.2,
+            );
+
+            let y_in = (y_ratio * h as f64).floor() as usize;
```
Comment (ashdtu) [inline crates/burn-ndarray/src/ops/interpolate.rs:39]: I think It's great that we are keeping this consistent with OpenCV version of this operation. Just as FYI, some methods may use ceil() or round() as well to find nearest coordinate so we may see Candle/Tch backend's implementation output slightly differ.  But i think this makes sense.

**Review by ashdtu — APPROVED** [review]
No specific correction comments. I think this looks great. The test cases cover both upsample, downsample operation. overall, LGTM!
